exercises/building_suffix_arrays/descr.md

```diff
@@ -0,0 +1,543 @@
+
+Manber and Myers' first insight is that the necessary comparisons *have
+already been performed* during previous stages! Indeed, suppose that two
+suffixes *d1* and *d2* lie in the same *h*-bucket. In order to determine in
+what order they should appear at the end of this stage, we need to
+*2h*-compare them, that is, to compare their *2h* first characters. However,
+we already know that their *h* first characters coincide. So, we need to
+compare the *next* run of *h* characters, between offsets *h* (included) and
+*2h* (excluded). In other words, we need to *h*-compare the suffixes *d1+h*
+and *d2+h*. (Let us assume in this informal explanation that *d1+h* and *d2+h*
+are both less than *n*, so they are well-defined suffixes.)
+This is something that we have already done during previous stages!
+Since the array *a* is *h*-sorted, all we need to do is find out which of the
+suffixes *d1+h* and *d2+h* appears first in the array *a*. With appropriate
+preprocessing (see following remark), this can be done in constant time.
+
+*Remark.* The permutation *a* can be inverted in time *O(n)*. Its inverse is
+an array that maps every suffix *d* to its position in the array *a*. Once
+this inverse permutation has been computed, we can find out in constant time
+where in the array *a* a suffix *d* is stored.
+
+To sum up, Manber and Myers' first insight is that two suffixes can be
+*2h*-compared in constant time.
+
+Thus, a naïve idea would be to implement stage *h* simply by sorting the array
+*a* using an off-the-shelf sorting algorithm, such as Heapsort or Mergesort.
+The total complexity of this stage would then be *O(n log n)*, which is not
+too bad.
+
+### Second Insight and Pseudocode
+
+Yet, one can do better! Manber and Myers' second insight is that it is
+possible to move every suffix to a correct position in time *O(n)*.
+
+To explain this, imagine that we keep the arrays *a* and *b* in their original
+state, and we create two new arrays *a'* and *b'* that describe the new state.
+(In practice, one may be able to re-use the existing storage instead of
+allocating new storage; this is an implementation detail.)
+
+From the array *a*, we will read all suffixes,
+from left to right, in *h*-order.
+Into the array *a'*, which is initially a copy of *a*,
+we will write all suffixes in the range *[0,n-h]*,
+in *2h*-order, at appropriate positions.
+Of course, we cannot write them all in a single
+left-to-right pass.
+Instead, we rely on the existing boundaries
+between *h*-buckets, which we keep.
+We think of each *h*-bucket as initially empty,
+so we view the array *a'* as a series of initially empty *h*-buckets.
+Inside each such *h*-bucket, we write from left to right.
+(This requires us to keep track, inside each *h*-bucket, of the
+boundary between the slots that have been written already and
+the slots that remain available.)
+
+At a high level of abstraction, the algorithm can be described as follows:
```

Comment:
If that's easy to do, you could show how each of the algorithm's steps performs on the "Mississippi" running example.

Reply:
I am not sure I am capable of doing this, and not sure it would be easy to follow... (I always have difficulty following animations, especially in the case of a complex algorithm like this.)